adapters/sonobi/sonobi.go

```diff
@@ -0,0 +1,296 @@
+		if err != nil {
+			return nil, err
+		}
+
+		// Fixes some segfaults. Since this is legacy code, I'm not looking into it too deeply
+		if len(sonobiReq.Imp) <= i {
+			break
+		}
+		sonobiReq.Imp[i].TagID = params.TagID
+	}
+
+	reqJSON, err := json.Marshal(sonobiReq)
+	if err != nil {
+		return nil, err
+	}
+
+	debug := &pbs.BidderDebug{
+		RequestURI: a.URI,
+	}
+	httpReq, _ := http.NewRequest("POST", a.URI, bytes.NewReader(reqJSON))
+	httpReq.Header.Set("Content-Type", "application/json")
+	if sReq.Device != nil {
+		addHeaderIfNonEmpty(httpReq.Header, "User-Agent", sReq.Device.UA)
+		addHeaderIfNonEmpty(httpReq.Header, "X-Forwarded-For", sReq.Device.IP)
+		addHeaderIfNonEmpty(httpReq.Header, "Accept-Language", sReq.Device.Language)
+		addHeaderIfNonEmpty(httpReq.Header, "DNT", strconv.Itoa(int(sReq.Device.DNT)))
+	}
+	sResp, err := ctxhttp.Do(ctx, a.http.Client, httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer sResp.Body.Close()
+	debug.StatusCode = sResp.StatusCode
+	if sResp.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	body, err := ioutil.ReadAll(sResp.Body)
+	if err != nil {
+		return nil, err
+	}
+	responseBody := string(body)
+	if sResp.StatusCode == http.StatusBadRequest {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+
+	if sResp.StatusCode != http.StatusOK {
+		return nil, &errortypes.BadServerResponse{
+			Message: fmt.Sprintf("HTTP status %d; body: %s", sResp.StatusCode, responseBody),
+		}
+	}
+	if req.IsDebug {
+		debug.RequestBody = string(reqJSON)
+		bidder.Debug = append(bidder.Debug, debug)
+		debug.ResponseBody = responseBody
+	}
+
+	var bidResp openrtb.BidResponse
+	err = json.Unmarshal(body, &bidResp)
+	if err != nil {
+		return nil, &errortypes.BadServerResponse{
+			Message: err.Error(),
+		}
+	}
+	bids := make(pbs.PBSBidSlice, 0)
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			bidID := bidder.LookupBidID(bid.ImpID)
+			if bidID == "" {
+				return nil, &errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}
+			}
+
+			adm, _ := url.QueryUnescape(bid.AdM)
+			pbid := pbs.PBSBid{
+				BidID:       bidID,
+				AdUnitCode:  bid.ImpID,
+				BidderCode:  bidder.BidderCode,
+				Price:       bid.Price,
+				Adm:         adm,
+				Creative_id: bid.CrID,
+				Width:       bid.W,
+				Height:      bid.H,
+				DealId:      bid.DealID,
+				NURL:        bid.NURL,
+			}
+			bids = append(bids, &pbid)
+		}
+	}
+
+	sort.Sort(bids)
+	return bids, nil
+}
+
+// MakeRequests Makes the OpenRTB request payload
+func (a *SonobiAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var sonobiExt openrtb_ext.ExtImpSonobi
+	var err error
+
+	var adapterRequests []*adapters.RequestData
+
+	for _, imp := range request.Imp {
+		// Sonobi doesn't allow multi-type imp. Banner takes priority over video.
```

Comment:
I think this means you want code someplace to set the video object to `nil` if the banner object exists. Or it may just be a note about what your bidder will do if it gets both in a bid request from PBS?

Reply:
I find this comment confusing, too.  It led me to expect that there should be some sort of code for the immediately following lines, where imp.Banner and imp.Video are non-nil.  Instead, it seems to simply refer to the fact that getMediaTypeForImp(), near the end of this file, uses banner if both banner and video are set. 

If there really is nothing to be done when imp.Banner and imp.Video are non-nil, it would be much clearer if the code either had comments indicating so (between lines 182/183 and between 183/184) or if the tests were rewritten to be "if imp.Banner == nil && imp.Video == nil".

Reply:
Must have been left over code from a while back when the first efforts to make this adapter was being made. I removed the check for media types since they will be filtered to the ones that Sonobi supports before hand.